File: app/src/main/java/com/dicoding/moviecataloguerv/views/detail/SeasonsActivity.java
package com.dicoding.moviecataloguerv.views.detail;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.dicoding.moviecataloguerv.adapter.EpisodeAdapter;
import com.dicoding.moviecataloguerv.databinding.ActivitySeasonsBinding;
import com.dicoding.moviecataloguerv.viewmodel.SeasonsViewModel;
import com.dicoding.moviecataloguerv.viewmodel.ViewModelFactory;

import java.util.ArrayList;

public class SeasonsActivity extends AppCompatActivity {

    public static String TV_SHOW_ID = "tvShow_id";
    public static String SEASON_NUMBER = "season_number";

    private int tvId;
    private int seasonNumber;

    private ActivitySeasonsBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySeasonsBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setSupportActionBar(binding.seasonToolbar);
        showLoading(true);

        tvId = getIntent().getIntExtra(TV_SHOW_ID, tvId);
        seasonNumber = getIntent().getIntExtra(SEASON_NUMBER, seasonNumber);

        EpisodeAdapter adapter = new EpisodeAdapter(new ArrayList<>());
        binding.rvEpisode.setLayoutManager(new LinearLayoutManager(this));
        binding.rvEpisode.setHasFixedSize(true);
        binding.rvEpisode.setAdapter(adapter);

        SeasonsViewModel viewModel = obtainViewModel(this);
        viewModel.getSeasons(tvId, seasonNumber).observe(this, episodes -> {
            if (episodes.getEpisodes() != null) {
                if (getSupportActionBar() != null) {
                    getSupportActionBar().setTitle(episodes.getName());
                    getSupportActionBar().setDisplayHomeAsUpEnabled(true);
                }
                adapter.setEpisodes(episodes.getEpisodes());
                showLoading(false);
            }
        });
    }

    @NonNull
    private static SeasonsViewModel obtainViewModel(AppCompatActivity activity) {
        ViewModelFactory factory = ViewModelFactory.getInstance(activity.getApplication());
        return new ViewModelProvider(activity, factory).get(SeasonsViewModel.class);
    }

    private void showLoading(boolean state) {
        if (state) {
            binding.progressBar.setVisibility(View.VISIBLE);
        } else {
            binding.progressBar.setVisibility(View.GONE);
        }
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return super.onSupportNavigateUp();
    }
}

File: app/src/main/kotlin/com/ryanrvldo/moviecatalogue/ui/detail/SeasonsActivity.kt
package com.ryanrvldo.moviecatalogue.ui.detail

import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import com.ryanrvldo.moviecatalogue.adapter.EpisodeAdapter
import com.ryanrvldo.moviecatalogue.data.model.Season
import com.ryanrvldo.moviecatalogue.databinding.ActivitySeasonsBinding
import com.ryanrvldo.moviecatalogue.ui.viewmodel.SeasonsViewModel
import com.ryanrvldo.moviecatalogue.utils.LayoutManagerUtil.getVerticalLayoutManager
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class SeasonsActivity : AppCompatActivity() {

    private lateinit var binding: ActivitySeasonsBinding
    private val viewModel: SeasonsViewModel by viewModels()

    @Inject
    lateinit var adapter: EpisodeAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySeasonsBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.seasonToolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        binding.rvEpisode.layoutManager = getVerticalLayoutManager(this)
        binding.rvEpisode.adapter = adapter

        viewModel.seasons.observe(this) { episodes: Season ->
            supportActionBar?.title = episodes.name
            adapter.differ.submitList(episodes.episodes)
        }
    }

    private fun showLoading(state: Boolean) {
        if (state) {
            binding.progressBar.visibility = View.VISIBLE
        } else {
            binding.progressBar.visibility = View.GONE
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return super.onSupportNavigateUp()
    }
}